enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/log/debug/DumpSegmentedRaftLog.java

```diff
@@ -0,0 +1,151 @@
+        int logsFound = 0;
+        RecoveryProtocol recoveryProtocol =
+                new RecoveryProtocol( fileSystem, new FileNames( new File( filenameOrDirectory ) ), marshal,
+                        logProvider );
+        Segments segments = recoveryProtocol.run().getSegments();
+
+        ListIterator<SegmentFile> segmentFileIterator = segments.getSegmentFileIteratorAtStart();
+
+        SegmentFile currentSegmentFile;
+        while (segmentFileIterator.hasNext())
+        {
+            currentSegmentFile = segmentFileIterator.next();
+            logsFound++;
+            out.println( "=== " + currentSegmentFile.getFilename() + " ===" );
+
+            SegmentHeader header = currentSegmentFile.header();
+
+            out.println( header.toString() );
+
+            try ( IOCursor<EntryRecord> cursor = currentSegmentFile.getReader( header.prevIndex() + 1 ) )
+            {
+                while ( cursor.next() )
+                {
+                    out.println( cursor.get().toString() );
+                }
+            }
+        }
+        return logsFound;
+    }
+
+    public static void main( String[] args ) throws IOException, DisposedException, DamagedLogStorageException
+    {
+        Args arguments = Args.withFlags( TO_FILE ).parse( args );
+        try ( Printer printer = getPrinter( arguments ) )
+        {
+            for ( String fileAsString : arguments.orphans() )
+            {
+                System.out.println( "Reading file " + fileAsString );
+                new DumpSegmentedRaftLog( new DefaultFileSystemAbstraction(), new CoreReplicatedContentMarshal() )
+                        .dump( fileAsString, printer.getFor( fileAsString ) );
+            }
+        }
+    }
+
+    public static Printer getPrinter( Args args )
+    {
+        boolean toFile = args.getBoolean( TO_FILE, false, true );
+        return toFile ? new DumpSegmentedRaftLog.FilePrinter() : SYSTEM_OUT_PRINTER;
+    }
+
+    public interface Printer extends AutoCloseable
+    {
+        PrintStream getFor( String file ) throws FileNotFoundException;
+
+        @Override
+        void close();
+    }
+
+    private static final Printer SYSTEM_OUT_PRINTER = new Printer()
+    {
+        @Override
+        public PrintStream getFor( String file )
+        {
+            return System.out;
+        }
+
+        @Override
+        public void close()
+        {   // Don't close System.out
+        }
+    };
+
+    private static class FilePrinter implements Printer
+    {
+        private File directory;
+        private PrintStream out;
+
+        @Override
+        public PrintStream getFor( String file ) throws FileNotFoundException
+        {
+            File absoluteFile = new File( file ).getAbsoluteFile();
+            File dir = absoluteFile.isDirectory() ? absoluteFile : absoluteFile.getParentFile();
+            if ( !dir.equals( directory ) )
+            {
+                safeClose();
```

Comment:
this a bit strange, isn't it more obvious to handle close on the PrintStream level instead? it shouldn't be the job of the a "factory". Yes, I know, we need to figure out a way to avoid closing System.out.